File: type-parser/test/testLib.hs
{-# LANGUAGE ForeignFunctionInterface #-}

module TestModule where

import Foreign
import Foreign.C
import Foreign.Ptr
import System.IO.Unsafe

import Tenkei

foreign import ccall "triple" foreign_triple :: Ptr Word8 -> CSize -> Ptr (Ptr Word8) -> Ptr CSize -> IO ()
triple :: Int32 -> Int32
triple = unsafePerformIO . (call foreign_triple)

foreign import ccall "quadruple" foreign_quadruple :: Ptr Word8 -> CSize -> Ptr (Ptr Word8) -> Ptr CSize -> IO ()
quadruple :: Int32 -> Int32
quadruple = unsafePerformIO . (call foreign_quadruple)

foreign import ccall "quintuple" foreign_quintuple :: Ptr Word8 -> CSize -> Ptr (Ptr Word8) -> Ptr CSize -> IO ()
quintuple :: Int32 -> Int32
quintuple = unsafePerformIO . (call foreign_quintuple)


data TestType = TestType { test :: Int32, otherTest :: Int32 }
data TestType2 = Test Int32 | OtherTest Int32

File: type-parser/test/testLib.rs
pub fn triple(x: i32) -> i32 {
    x * 3
}

pub fn quadruple(x: i32) -> i32 {
    x * 4
}

pub fn quintuple(x: i32) -> i32 {
    x * 5
}

pub struct TestType {
    test: i32,
    other_test: i32,
}

pub enum TestType2 {
    Test(i32),
    OtherTest(i32),
}
